icu4j/main/classes/core/src/com/ibm/icu/impl/personname/PersonNameFormatterImpl.java

```diff
@@ -215,32 +224,84 @@ private String[] asStringArray(ICUResourceBundle resource) {
+    }
+}
+
+    private static String closestMatchingLocaleID(Locale locale, Set<String> localesToMatch) {
+        // This function is called by nameIsGnFirst() to find the most specific match for a locale's
+        // ID in a set of locale IDs (which will always be the union of gnFirstLocales and snFirstLocales--
+        // that's passed in so that we don't calculate it over and over again). We take the given
+        // locale ID and then walk its parent chain until we find something in localesToMatch that matches it
+        // (localesToMatch has to contain "und", and "und" is always at the end of the chain).  At each step,
+        // we try both the locale as given and the locale with "und" substituted for the language code,
+        // so that things like "zh_CN" can match things like "und_CN".
+        if (locale == null) {
+            return null;
+        }
+
+        // this is a hack to deal with certain script codes that are valid, but not the default, for their locales--
+        // to make the parent-chain lookup work right, we need to replace any of those script codes  with the appropriate
+        // default script for whatever language and region we have
+        ULocale uLocale = ULocale.forLocale(locale);
+        List<String> NON_DEFAULT_SCRIPTS = Arrays.asList(new String[] { "Hani", "Hira", "Kana" });
+        if (NON_DEFAULT_SCRIPTS.contains(uLocale.getScript())) {
+            ULocale.Builder builder = new ULocale.Builder();
+            builder.setLocale(uLocale);
+            builder.setScript(null);
+            uLocale = ULocale.addLikelySubtags(builder.build());
         }
 
-        String localeStr = getNameLocale(name).toString();
+        String localeStr = uLocale.getName();
+        String origLocaleStr = localeStr;
+        String languageCode = uLocale.getLanguage();
+
         do {
-            if (gnFirstLocales.contains(localeStr)) {
-                return true;
-            } else if (snFirstLocales.contains(localeStr)) {
-                return false;
+            if (localesToMatch.contains(localeStr)) {
+                return localeStr;
             }
 
-            int lastUnderbarPos = localeStr.lastIndexOf("_");
-            if (lastUnderbarPos >= 0) {
-                localeStr = localeStr.substring(0, lastUnderbarPos);
-            } else {
-                localeStr = "root";
+            // before looking up the parent locale, try again with "und" in place of the language code
+            String undStr = localeStr.replaceAll("^" + languageCode, "und");
+            if (localesToMatch.contains(undStr)) {
+                return undStr;
             }
-        } while (!localeStr.equals("root"));
 
-        // should never get here-- "root" should always be in one of the locales
-        return true;
+            String parentLocaleStr = ICUResourceBundle.getParentLocaleID(localeStr, origLocaleStr, ICUResourceBundle.OpenType.LOCALE_DEFAULT_ROOT);
```

Comment:
I don't like calling `ICUResourceBundle` like this, but I'm uncomfortable moving that logic into `ULocale`, where it probably belongs...

Reply:
That is really ugly... Didn't realize that the functionality was just in the RBs! @markusicu was this in case applications wanted their own parent relationship?

Reply:
I think I may have actually added that function-- I know I added to it-- when I was making all the horizontal-inheritance fixes a while back.  And at that point, the only thing that needed it was ICUResourceBundle.  I'm still worried, though, that it's not quite ready to put into ULocale, at least if that would mean making it public.

One of the things that made my logic here complicated was that that method returns null if you change scripts moving up the parent chain, but I think we actually do want to do that here...

Reply:
What should happen is you skip from (say) en-Cyrl to root. So you only change scripts if the script is **not** the default for the language. And after root in any case you should fail (return null).